DFC.ServiceTaxonomy.Neo4j/Services/NeoGraphDatabase.cs

```diff
@@ -1,42 +1,51 @@
 using System;
 using System.Collections.Generic;
+using System.Linq;
 using System.Threading.Tasks;
+using DFC.ServiceTaxonomy.Neo4j.Commands.Interfaces;
 using DFC.ServiceTaxonomy.Neo4j.Configuration;
+using DFC.ServiceTaxonomy.Neo4j.Queries.Interfaces;
 using Microsoft.Extensions.Options;
+using Microsoft.Extensions.Logging;
 using Neo4j.Driver;
+using ILogger = Neo4j.Driver.ILogger;
 
 namespace DFC.ServiceTaxonomy.Neo4j.Services
 {
-    //todo: ensure sync to neo when importing data through recipe
-
     // https://github.com/neo4j/neo4j-dotnet-driver
-    // (not updated for 4 yet: https://neo4j.com/docs/driver-manual/1.7/get-started/)
+    // https://neo4j.com/docs/driver-manual/4.0/
     public class NeoGraphDatabase : IGraphDatabase, IDisposable
     {
+        private readonly ILogger<NeoGraphDatabase> _logger;
         private readonly IDriver _driver;
 
-        public NeoGraphDatabase(IOptionsMonitor<Neo4jConfiguration> neo4jConfigurationOptions)
+        public NeoGraphDatabase(
+            IOptionsMonitor<Neo4jConfiguration> neo4jConfigurationOptions,
+            ILogger neoLogger,
+            ILogger<NeoGraphDatabase> logger)
         {
+            _logger = logger;
             // Each IDriver instance maintains a pool of connections inside, as a result, it is recommended to only use one driver per application.
             // It is considerably cheap to create new sessions and transactions, as sessions and transactions do not create new connections as long as there are free connections available in the connection pool.
             //  driver is thread-safe, while the session or the transaction is not thread-safe.
             //todo: add configuration/settings menu item/page so user can enter this
             var neo4jConfiguration = neo4jConfigurationOptions.CurrentValue;
 
-            //todo: pass logger, see https://github.com/neo4j/neo4j-dotnet-driver
-            // o => o.WithLogger(logger)
-            _driver = GraphDatabase.Driver(neo4jConfiguration.Endpoint.Uri, AuthTokens.Basic(neo4jConfiguration.Endpoint.Username, neo4jConfiguration.Endpoint.Password));
+            _driver = GraphDatabase.Driver(
+                neo4jConfiguration.Endpoint.Uri,
+                AuthTokens.Basic(neo4jConfiguration.Endpoint.Username, neo4jConfiguration.Endpoint.Password),
+                o => o.WithLogger(neoLogger));
         }
 
-        public async Task<List<T>> RunReadQuery<T>(Query query, Func<IRecord, T> operation)
+        public async Task<List<T>> RunReadQuery<T>(IQuery<T> query)
         {
             IAsyncSession session = _driver.AsyncSession();
             try
             {
                 return await session.ReadTransactionAsync(async tx =>
```

Comment:
If the read or write transactions throw an exception, is it worth logging in the finally clauses?

Reply:
Any exceptions thrown within the transaction, get logged by Neo/NeoLogger, e.g. if a command validation fails, we see this in the logfile...
```
DFC.ServiceTaxonomy.Neo4j.Log.NeoLogger|ERROR|[null] DFC.ServiceTaxonomy.Neo4j.Exceptions.CommandValidationException: Expecting 1 node to be deleted, but 0 were actually deleted.
```